File: app/src/main/java/ch/threema/app/ui/TooltipPopup.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2018-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.ui;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.BitmapDrawable;
import android.os.Handler;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.PopupWindow;

import androidx.annotation.DrawableRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.preference.PreferenceManager;

import com.google.android.material.card.MaterialCardView;

import ch.threema.app.R;
import ch.threema.app.emojis.EmojiTextView;
import ch.threema.app.utils.ConfigUtils;

public class TooltipPopup extends PopupWindow implements DefaultLifecycleObserver {

	public static final int ALIGN_ABOVE_ANCHOR_ARROW_LEFT = 1;
	public static final int ALIGN_BELOW_ANCHOR_ARROW_RIGHT = 2;
	public static final int ALIGN_BELOW_ANCHOR_ARROW_LEFT = 3;
	public static final int ALIGN_ABOVE_ANCHOR_ARROW_RIGHT = 4;

	private final Context context;
	private View popupLayout;
	private EmojiTextView textView;
	private final String preferenceString;
	private Handler timeoutHandler;
	private final Runnable dismissRunnable = () -> dismiss(false);

	private @DrawableRes int icon = 0;

	public TooltipPopup(Context context, int preferenceKey, LifecycleOwner lifecycleOwner) {
		super(context);

		if (lifecycleOwner != null) {
			lifecycleOwner.getLifecycle().addObserver(this);
		}

		if (preferenceKey == 0) {
			this.preferenceString = null;
		} else {
			this.preferenceString = context.getString(preferenceKey);
		}
		this.context = context;

		init(context, null);
	}

	public TooltipPopup(Context context, int preferenceKey, LifecycleOwner lifecycleOwner, Intent launchIntent, @DrawableRes int icon) {
		super(context);

		if (lifecycleOwner != null) {
			lifecycleOwner.getLifecycle().addObserver(this);
		}

		if (preferenceKey == 0) {
			this.preferenceString = null;
		} else {
			this.preferenceString = context.getString(preferenceKey);
		}
		this.context = context;
		this.icon = icon;

		init(context, launchIntent);
	}

	private void init(Context context, Intent launchIntent) {
		LayoutInflater layoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

		this.popupLayout = layoutInflater.inflate(R.layout.popup_tooltip, null, false);
		this.textView = popupLayout.findViewById(R.id.label);

		setContentView(popupLayout);
		setInputMethodMode(PopupWindow.INPUT_METHOD_NOT_NEEDED);
		setAnimationStyle(R.style.TooltipAnimation);
		setFocusable(false);
		setTouchable(true);
		setOutsideTouchable(false);
		setBackgroundDrawable(new BitmapDrawable());

		popupLayout.setOnClickListener(v -> {
			if (launchIntent != null) {
				context.startActivity(launchIntent);
				if (context instanceof Activity) {
					((Activity) context).overridePendingTransition(0, 0);
				}
			} else {
				dismissForever();
			}
		});

		View closeButton = popupLayout.findViewById(R.id.close_button);
		if (closeButton != null) {
			if (preferenceString == null) {
				closeButton.setVisibility(View.GONE);
			} else {
				closeButton.setOnClickListener(v -> dismissForever());
			}
		}
	}

	public static boolean isDismissed(Context context, String preferenceString) {
		if (preferenceString == null) {
			return false;
		}

		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
		if (sharedPreferences != null) {
			return sharedPreferences.getBoolean(preferenceString, false);
		}
		return false;
	}

	public void dismissForever() {
		if (preferenceString != null) {
			SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
			if (sharedPreferences != null) {
				sharedPreferences.edit().putBoolean(preferenceString, true).apply();
			}
		}

		dismiss(false);
	}

	public void dismiss(boolean immediate) {
		if (immediate) {
			setAnimationStyle(0);
		}

		if (timeoutHandler != null) {
			timeoutHandler.removeCallbacks(dismissRunnable);
			timeoutHandler = null;
		}

		this.dismiss();
	}

	/**
	 * Show a tooltip at the specified location pointing to a specified anchor view
	 * @param activity Activity context
	 * @param anchor Anchor / parent view to of this tooltip
	 * @param text Text to show in tooltip
	 * @param align Where to align the tooltip and where the arrow should be shown
	 * @param originLocation The location on screen where the tip of the arrow should point to
	 * @param timeoutMs How long the tooltip should be shown until it fades out
	 */
	public void show(Activity activity, final View anchor, String text, int align, int[] originLocation, int timeoutMs) {
		if (isDismissed(context, preferenceString)) {
			return;
		}

		this.textView.setText(text);

		int screenHeight = activity.getWindowManager().getDefaultDisplay().getHeight();
		int screenWidth = activity.getWindowManager().getDefaultDisplay().getWidth();
		int maxWidth = context.getResources().getDimensionPixelSize(R.dimen.tooltip_max_width);
		int arrowInset = context.getResources().getDimensionPixelSize(R.dimen.tooltip_popup_arrow_inset);
		int marginOnOtherEdge = context.getResources().getDimensionPixelSize(R.dimen.tooltip_margin_on_other_edge);
		int arrowOffset = (context.getResources().getDimensionPixelSize(R.dimen.identity_popup_arrow_width) / 2) + arrowInset;
		int popupX, popupY, popupWidth, anchorGravity, contentGravity;

		if (align == ALIGN_ABOVE_ANCHOR_ARROW_LEFT) {
			this.popupLayout.findViewById(R.id.arrow_bottom_left).setVisibility(View.VISIBLE);
			popupX = Math.max(0, originLocation[0] - arrowOffset); // left edge of popup
			popupY = screenHeight - originLocation[1] + ConfigUtils.getNavigationBarHeight(activity);
			popupWidth = Math.min(screenWidth - popupX - marginOnOtherEdge, maxWidth);
			anchorGravity = Gravity.LEFT | Gravity.BOTTOM;
			contentGravity = Gravity.LEFT;
		} else if (align == ALIGN_ABOVE_ANCHOR_ARROW_RIGHT) {
			this.popupLayout.findViewById(R.id.arrow_bottom_right).setVisibility(View.VISIBLE);
			popupX = Math.min(screenWidth, originLocation[0] + arrowOffset);
			popupY = screenHeight - originLocation[1] + ConfigUtils.getNavigationBarHeight(activity);
			popupWidth = Math.min(popupX - marginOnOtherEdge, maxWidth);
			popupX -= popupWidth;
			anchorGravity = Gravity.LEFT | Gravity.BOTTOM;
			contentGravity = Gravity.RIGHT;
		} else if (align == ALIGN_BELOW_ANCHOR_ARROW_LEFT) {
			this.popupLayout.findViewById(R.id.arrow_top_left).setVisibility(View.VISIBLE);
			popupX = Math.max(0, originLocation[0] - arrowOffset); // left edge of popup
			popupY = originLocation[1];
			popupWidth = Math.min(screenWidth - popupX - marginOnOtherEdge, maxWidth);
			anchorGravity = Gravity.LEFT | Gravity.TOP;
			contentGravity = Gravity.LEFT;
		} else { // arrow right
			this.popupLayout.findViewById(R.id.arrow_top_right).setVisibility(View.VISIBLE);
			popupX = Math.min(screenWidth, originLocation[0] + arrowOffset); // right edge of popup
			popupY = originLocation[1];
			popupWidth = Math.min(popupX - marginOnOtherEdge, maxWidth);
			popupX -= popupWidth;
			anchorGravity = Gravity.LEFT | Gravity.TOP;
			contentGravity = Gravity.RIGHT;
		}

		this.setWidth(popupWidth);
		MaterialCardView contentLayout = this.popupLayout.findViewById(R.id.content);
		FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) contentLayout.getLayoutParams();
		params.gravity = contentGravity;
		contentLayout.setLayoutParams(params);

		if (activity.isFinishing() || activity.isDestroyed()) {
			return;
		}
		try {
			showAtLocation(anchor, anchorGravity, popupX, popupY);
		} catch (WindowManager.BadTokenException e) {
			return;
		}

		ImageView iconView = this.popupLayout.findViewById(R.id.icon);
		if (icon != 0) {
			iconView.setImageResource(icon);
			iconView.setVisibility(View.VISIBLE);
		} else {
			iconView.setVisibility(View.GONE);
		}

		if (timeoutMs > 0) {
			if (timeoutHandler == null) {
				timeoutHandler = new Handler();
			}
			timeoutHandler.removeCallbacks(dismissRunnable);
			timeoutHandler.postDelayed(dismissRunnable, timeoutMs);
		}
	}

	/**
	 * Notifies that {@code ON_PAUSE} event occurred.
	 * <p>
	 * This method will be called before the {@link LifecycleOwner}'s {@code onPause} method
	 * is called.
	 *
	 * @param owner the component, whose state was changed
	 */
	@Override
	public void onPause(@NonNull LifecycleOwner owner) {
		dismiss(true);
	}
}

File: app/src/main/java/ch/threema/app/ui/TooltipPopup.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2018-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.app.ui

import android.app.Activity
import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager.BadTokenException
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.PopupWindow
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.core.content.edit
import androidx.core.content.getSystemService
import androidx.core.view.isVisible
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.preference.PreferenceManager
import ch.threema.app.R
import ch.threema.app.emojis.EmojiTextView
import ch.threema.app.utils.ConfigUtils
import com.google.android.material.card.MaterialCardView


class TooltipPopup
@JvmOverloads
constructor(
    private val context: Context,
    @StringRes
    preferenceKey: Int = 0,
    lifecycleOwner: LifecycleOwner? = null,
    @DrawableRes
    private val icon: Int = 0,
    private val showCloseButton: Boolean = true,
) : PopupWindow(context), DefaultLifecycleObserver {
    private var popupLayout: View
    private var titleView: EmojiTextView
    private var textView: EmojiTextView
    private var preferenceString: String? = null
    private var timeoutHandler: Handler? = null
    private val dismissRunnable = Runnable {
        listener.onTimedOut(this)
        dismiss(false)
    }

    var listener: TooltipPopupListener = TooltipPopupListener()

    init {
        lifecycleOwner?.lifecycle?.addObserver(this)
        preferenceString = if (preferenceKey != 0) {
            context.getString(preferenceKey)
        } else {
            null
        }

        val layoutInflater = context.getSystemService<LayoutInflater>()!!
        popupLayout = layoutInflater.inflate(R.layout.popup_tooltip, null, false)!!
        titleView = popupLayout.findViewById(R.id.title)
        textView = popupLayout.findViewById(R.id.label)
        contentView = popupLayout
        inputMethodMode = INPUT_METHOD_NOT_NEEDED
        animationStyle = R.style.TooltipAnimation
        isFocusable = false
        isTouchable = true
        isOutsideTouchable = false
        setBackgroundDrawable(ColorDrawable())

        popupLayout.setOnClickListener {
            listener.onClicked(this)
        }

        popupLayout.findViewById<View>(R.id.close_button)
            ?.let { closeButton ->
                closeButton.isVisible = showCloseButton
                closeButton.setOnClickListener {
                    listener.onCloseButtonClicked(this)
                }
            }
    }

    fun dismissForever() {
        if (preferenceString != null) {
            PreferenceManager.getDefaultSharedPreferences(context)
                ?.edit {
                    putBoolean(preferenceString, true)
                }
        }

        dismiss(false)
    }

    fun dismiss(immediate: Boolean) {
        if (immediate) {
            animationStyle = 0
        }

        timeoutHandler?.removeCallbacks(dismissRunnable)
        timeoutHandler = null

        this.dismiss()
        listener.onDismissed(this)
    }

    /**
     * Show a tooltip at the specified location pointing to a specified anchor view
     *
     * @param activity       Activity context
     * @param anchor         Anchor / parent view to of this tooltip
     * @param title          Optional title text to show in tooltip
     * @param text           Text to show in tooltip
     * @param alignment         Where to align the tooltip and where the arrow should be shown
     * @param originLocation The location on screen where the tip of the arrow should point to
     * @param timeoutMs      How long the tooltip should be shown until it fades out
     */
    fun show(
        activity: Activity,
        anchor: View,
        title: String? = null,
        text: String?,
        alignment: Alignment,
        originLocation: IntArray,
        timeoutMs: Int = 0,
    ) {
        if (isForeverDismissed()) {
            return
        }

        if (title != null) {
            titleView.text = title
            titleView.isVisible = true
        } else {
            titleView.isVisible = false
        }
        textView.text = text

        val resources = context.resources

        val screenHeight: Int
        val screenWidth: Int
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val windowMetrics = activity.windowManager.currentWindowMetrics
            val rect = windowMetrics.bounds
            screenWidth = rect.right
            screenHeight = rect.bottom
        } else {
            screenWidth = resources.displayMetrics.widthPixels
            screenHeight = resources.displayMetrics.heightPixels + ConfigUtils.getNavigationBarHeight(activity)
        }
        val maxWidth = resources.getDimensionPixelSize(R.dimen.tooltip_max_width)
        val arrowInset = resources.getDimensionPixelSize(R.dimen.tooltip_popup_arrow_inset)
        val marginOnOtherEdge = resources.getDimensionPixelSize(R.dimen.tooltip_margin_on_other_edge)
        val arrowOffset = (resources.getDimensionPixelSize(R.dimen.identity_popup_arrow_width) / 2) + arrowInset
        var popupX: Int
        val popupY: Int
        val popupWidth: Int
        val anchorGravity: Int
        val contentGravity: Int

        when (alignment) {
            Alignment.ABOVE_ANCHOR_ARROW_LEFT -> {
                popupLayout.findViewById<View>(R.id.arrow_bottom_left).isVisible = true
                popupX = (originLocation[0] - arrowOffset).coerceAtLeast(0)
                popupY = screenHeight - originLocation[1]
                popupWidth = (screenWidth - popupX - marginOnOtherEdge).coerceAtMost(maxWidth)
                anchorGravity = Gravity.LEFT or Gravity.BOTTOM
                contentGravity = Gravity.LEFT
            }

            Alignment.ABOVE_ANCHOR_ARROW_RIGHT -> {
                popupLayout.findViewById<View>(R.id.arrow_bottom_right).isVisible = true
                popupX = (originLocation[0] + arrowOffset).coerceAtMost(screenWidth)
                popupY = screenHeight - originLocation[1]
                popupWidth = (popupX - marginOnOtherEdge).coerceAtMost(maxWidth)
                popupX -= popupWidth
                anchorGravity = Gravity.LEFT or Gravity.BOTTOM
                contentGravity = Gravity.RIGHT
            }

            Alignment.BELOW_ANCHOR_ARROW_LEFT -> {
                popupLayout.findViewById<View>(R.id.arrow_top_left).isVisible = true
                popupX = (originLocation[0] - arrowOffset).coerceAtLeast(0)
                popupY = originLocation[1]
                popupWidth = (screenWidth - popupX - marginOnOtherEdge).coerceAtMost(maxWidth)
                anchorGravity = Gravity.LEFT or Gravity.TOP
                contentGravity = Gravity.LEFT
            }

            Alignment.BELOW_ANCHOR_ARROW_RIGHT -> {
                popupLayout.findViewById<View>(R.id.arrow_top_right).isVisible = true
                popupX = (originLocation[0] + arrowOffset).coerceAtMost(screenWidth)
                popupY = originLocation[1]
                popupWidth = (popupX - marginOnOtherEdge).coerceAtMost(maxWidth)
                popupX -= popupWidth
                anchorGravity = Gravity.LEFT or Gravity.TOP
                contentGravity = Gravity.RIGHT
            }
        }

        this.width = popupWidth
        this.height = ViewGroup.LayoutParams.WRAP_CONTENT
        val contentLayout = popupLayout.findViewById<MaterialCardView>(R.id.content)
        val params = contentLayout.layoutParams as FrameLayout.LayoutParams
        params.gravity = contentGravity
        contentLayout.layoutParams = params

        if (activity.isFinishing || activity.isDestroyed) {
            return
        }
        try {
            showAtLocation(anchor, anchorGravity, popupX, popupY)
        } catch (e: BadTokenException) {
            return
        }

        val iconView = popupLayout.findViewById<ImageView>(R.id.icon)
        if (icon != 0) {
            iconView.setImageResource(icon)
            iconView.isVisible = true
        } else {
            iconView.isVisible = false
        }

        listener.onShown(this)

        if (timeoutMs > 0) {
            if (timeoutHandler == null) {
                timeoutHandler = Handler(Looper.getMainLooper())
            }
            timeoutHandler?.removeCallbacks(dismissRunnable)
            timeoutHandler?.postDelayed(dismissRunnable, timeoutMs.toLong())
        }
    }

    private fun isForeverDismissed(): Boolean {
        if (preferenceString == null) {
            return false
        }
        return PreferenceManager.getDefaultSharedPreferences(context)
            ?.getBoolean(preferenceString, false)
            ?: false
    }

    /**
     * Notifies that `ON_PAUSE` event occurred.
     *
     * This method will be called before the [LifecycleOwner]'s `onPause` method
     * is called.
     *
     * @param owner the component, whose state was changed
     */
    override fun onPause(owner: LifecycleOwner) {
        dismiss(true)
    }

    open class TooltipPopupListener {
        open fun onShown(tooltipPopup: TooltipPopup) {}
        open fun onClicked(tooltipPopup: TooltipPopup) {
            tooltipPopup.dismissForever()
        }

        open fun onCloseButtonClicked(tooltipPopup: TooltipPopup) {
            tooltipPopup.dismissForever()
        }

        open fun onTimedOut(tooltipPopup: TooltipPopup) {}
        open fun onDismissed(tooltipPopup: TooltipPopup) {}
    }

    enum class Alignment {
        ABOVE_ANCHOR_ARROW_LEFT,
        ABOVE_ANCHOR_ARROW_RIGHT,
        BELOW_ANCHOR_ARROW_LEFT,
        BELOW_ANCHOR_ARROW_RIGHT,
    }
}
